src/org/rascalmpl/uri/URIUtil.java

```diff
@@ -433,4 +433,32 @@ public static ISourceLocation removeOffset(ISourceLocation prev) {
     public static ISourceLocation createFromURI(String value) throws URISyntaxException {
         return vf.sourceLocation(createFromEncoded(value));
     }
+    public static ISourceLocation changeExtension(ISourceLocation location, String ext) throws URISyntaxException {
+        String path = location.getPath();
+		boolean endsWithSlash = path.endsWith(URIUtil.URI_PATH_SEPARATOR);
+		if (endsWithSlash) {
+			path = path.substring(0, path.length() - 1);
```

Comment:
We have a semantics for "filenames" that includes directory names. A filename is the last segment of the path list. If a URI ends with a `/` we interpret the characters between that slash and the previous as the "filename", because that is the name of a directory in that case. Filenames can have extensions, and so for the sake of consistency, directory names can also have extensions. From there on you get to write this kind of code to make it all consistent.

All of this was discovered some years ago while writing simple random tests for the field update feature of Rascal. If you update a file name that does not have an extension, and then ask back for the extension, (or the new filename), then you want to see the extension back (or the filename with the extension) and not "hello/.txt" or the ".txt" for the new filename, or even "", just because the URI accidentally was a directory.